kernel: add nft tproxy support

shenwii opened the pull request:
Added kmod-nft-tproxy package to enable nft to support tproxy.

tiagogaspar8 commented on package/kernel/linux/modules/netfilter.mk:
```diff
@@ -1190,3 +1190,14 @@ define KernelPackage/nft-compat
+define KernelPackage/nft-tproxy
+  SUBMENU:=$(NF_MENU)
+  TITLE:=Netfilter nf_tables tproxy support
+  DEPENDS:=+kmod-nft-core kmod-ipt-tproxy
```
Based on:
https://cateee.net/lkddb/web-lkddb/NFT_TPROXY.html
There is no dependency on ipt-tproxy, why did you add that? I believe it should be the other way around.

shenwii replied:
```
config NFT_TPROXY
	tristate "Netfilter nf_tables tproxy support"
	depends on IPV6 || IPV6=n
	select NF_DEFRAG_IPV4
	select NF_DEFRAG_IPV6 if NF_TABLES_IPV6
	select NF_TPROXY_IPV4
	select NF_TPROXY_IPV6 if NF_TABLES_IPV6
	help
	  This makes transparent proxy support available in nftables.
```
Kconfig shows that NFT_TPROXY depends on: NF_DEFRAG_IPV4, NF_DEFRAG_IPV6, NF_TPROXY_IPV4, NF_TPROXY_IPV6.
```
# tproxy

$(eval $(call nf_add,IPT_TPROXY,CONFIG_NETFILTER_XT_MATCH_SOCKET, $(P_XT)xt_socket))
$(eval $(call nf_add,IPT_TPROXY,CONFIG_NF_SOCKET_IPV4, $(P_V4)nf_socket_ipv4))
$(eval $(call nf_add,IPT_TPROXY,CONFIG_NF_SOCKET_IPV6, $(P_V6)nf_socket_ipv6))
$(eval $(call nf_add,IPT_TPROXY,CONFIG_NETFILTER_XT_TARGET_TPROXY, $(P_XT)xt_TPROXY))
$(eval $(call nf_add,IPT_TPROXY,CONFIG_NF_TPROXY_IPV4, $(P_V4)nf_tproxy_ipv4))
$(eval $(call nf_add,IPT_TPROXY,CONFIG_NF_TPROXY_IPV6, $(P_V6)nf_tproxy_ipv6))
```
Then I found them in openwrt's IPT_TPROXY definition. So I added ipt-tproxy to the nft-tproxy dependency.

shenwii replied:
```
```
$(eval $(if $(NF_KMOD),$(call nf_add,NF_CONNTRACK,CONFIG_NF_DEFRAG_IPV6, $(P_V6)nf_defrag_ipv6),))
```
# kernel only
$(eval $(if $(NF_KMOD),$(call nf_add,NF_CONNTRACK,CONFIG_NF_CONNTRACK, $(P_XT)nf_conntrack),))
$(eval $(if $(NF_KMOD),$(call nf_add,NF_CONNTRACK,CONFIG_NF_DEFRAG_IPV4, $(P_V4)nf_defrag_ipv4),))
```
Addendum: ipt-tproxy depends on ipt-conntrack, and ipt-conntrack depends on nf-conntrack.

tiagogaspar8 replied:
Yes, but that dependency pulls in iptables/xtables which is to be dropped soon.
I believe you should add those configs that you need to the nf-tproxy package you're adding and removing them from the ipt-tproxy, then you make ipt-tproxy depend on nf-tproxy.
Tomorrow when I'm at the computer I can give you some more pointer if you need, but I believe you get it.
Just out of curiosity, do you need this for some package?

tiagogaspar8 replied:
Yes, but that dependency pulls in iptables/xtables which is to be dropped soon.
I believe you should add those configs that you need to the nf-tproxy package you're adding and removing them from the ipt-tproxy, then you make ipt-tproxy depend on nf-tproxy.
Tomorrow when I'm at the computer I can give you some more pointer if you need, but I believe you get it.
Just out of curiosity, do you need this for some package?

Update to your addendum:
Yeah, you should chenge the dependency to nf_conntrack and that way you pull in what you need without selecting iptables

shenwii replied:
> Just out of curiosity, do you need this for some package?

I'm a developer coming from China and the Chinese network you know, can't even access Google, which greatly affects my work.
PS: Baidu never returns the results I want.

In the past, I have used iptables tproxy on my router to redirect packets to proxy. After recently upgrading to the latest openwrt, I found that nft replaced iptables, but nft's tproxy doesn't work. Now I can only use nat to redirect ipv4 tcp packets, ipv6 and udp don't work.

Then I asked for help on the openwrt forum, but no one replied to me.
[https://forum.openwrt.org/t/how-to-make-nft-support-tproxy/118518](url)
[https://forum.openwrt.org/t/request-to-compile-nft-tproxy-ko-to-make-nft-support-tproxy/119974](url)

So I had to learn how to compile openwrt and how to add packages to openwrt by myself.

shenwii replied:
Thanks for the suggestion, I've readjusted the package dependencies, please review again.

PS: I found that the first log was not written according to the specification, so I reworked it.

tiagogaspar8 replied:
Yeah, I believe no one uses these modules, but your work is greatly appreciated!

tiagogaspar8 replied:
Yeah, I believe no one uses these modules, but your work is greatly appreciated!

mans0n commented:
Hi, this has some formal issues, especially:
* Please add a proper commit message to the first commit.
* Please add a Signed-off-by tag with your real name.
https://openwrt.org/docs/guide-developer/frequent-pr-mistakes

tiagogaspar8 commented on include/netfilter.mk:
```diff
@@ -241,12 +241,16 @@ $(eval $(call nf_add,IPT_DEBUG,CONFIG_NETFILTER_XT_TARGET_TRACE, $(P_XT)xt_TRACE
+$(eval $(if $(NF_KMOD),$(call nf_add,NF_TPROXY,CONFIG_NF_TPROXY_IPV4, $(P_V4)nf_tproxy_ipv4),))
+$(eval $(if $(NF_KMOD),$(call nf_add,NF_TPROXY,CONFIG_NF_TPROXY_IPV6, $(P_V6)nf_tproxy_ipv6),))
+$(eval $(if $(NF_KMOD),$(call nf_add,NF_TPROXY,CONFIG_NF_SOCKET_IPV4, $(P_V4)nf_socket_ipv4),))
+$(eval $(if $(NF_KMOD),$(call nf_add,NF_TPROXY,CONFIG_NF_SOCKET_IPV6, $(P_V6)nf_socket_ipv6),))
```
Why did you select CONFIG_NF_SOCKET_IPV6, I don't believe this is necessary for tproxy (neither for ipt_tproxy but that's an issue for the next guy 😉 )

shenwii commented on include/netfilter.mk:
```diff
@@ -241,12 +241,16 @@ $(eval $(call nf_add,IPT_DEBUG,CONFIG_NETFILTER_XT_TARGET_TRACE, $(P_XT)xt_TRACE
+
 $(eval $(call nf_add,IPT_TPROXY,CONFIG_NETFILTER_XT_MATCH_SOCKET, $(P_XT)xt_socket))
-$(eval $(call nf_add,IPT_TPROXY,CONFIG_NF_SOCKET_IPV4, $(P_V4)nf_socket_ipv4))
-$(eval $(call nf_add,IPT_TPROXY,CONFIG_NF_SOCKET_IPV6, $(P_V6)nf_socket_ipv6))
```
I didn't intend to add nf socket at first, but I was curious why ipt-tproxy added it, so I googled it and I found this article.

[kernel networking tproxy](https://www.kernel.org/doc/html/latest/networking/tproxy.html)

I think this is the reason why nf socket was added to ipt-tproxy before, so I chose to stay with the previous developers.

tiagogaspar8 replied:
I see but in that case you should add them as separate (separating the ipv4 and ipv6 versions):
- nf-socket: with CONFIG_NF_SOCKET
- nf-defrag: with CONFIG_NF_DEFRAG
- nf-conntrack: you need to add the dependency to nf-defrag
- nf-tproxy: with CONFIG_NF_TPROXY like you did but this time with no depends
- nft-tproxy: depending on nf-tproxy and nft-core
- nft-socket: depending on nf-socket

shenwii replied:
Thanks for the help.

I have a question about IPv6, originally nf-conntrack packet contains nf_defrag and nf_defrag6, but nf-conntrack6 packet is empty, why is that?
I also put nf_defrag and nf_defrag6 into the nf-defrag package to keep it the same as the original one, while nf-defrag6 doesn't put anything, I don't know if this is a good idea.

Sorry for my poor English, I don't know if I have expressed the meaning clearly.

tiagogaspar8 replied:
> I have a question about IPv6, originally nf-conntrack packet contains nf_defrag and nf_defrag6, but nf-conntrack6 packet is empty, why is that?

That is indeed a mistake, that must be corrected.

> I also put nf_defrag and nf_defrag6 into the nf-defrag package to keep it the same as the original one, while nf-defrag6 doesn't put anything, I don't know if this is a good idea.

This I did not understand, can you try to explain in another way please? 😃

shenwii replied:
> Here I believe you forgot to add the "6"

Please refer to the reply of this

Commit:
kernel: add nft tproxy support

add kmod-nft-tproxy package to make nft support tproxy

Signed-off-by: Shen Wei <654123789sw@gmail.com>

Commit:
kernel: add kmod-nf-tproxy package and adjusted dependencies

removed nf_tproxy_ipv4 and nf_tproxy_ipv6 from the kmod-ipt-tproxy package
and added them to the kmod-nf-tproxy package

made kmod-ipt-tproxy dependent on kmod-nf-tproxy and kmod-nft-tproxy
dependent on kmod-nf-tproxy, so that kmod-nft-tproxy no longer depends on
any iptables/xtables module

Signed-off-by: Shen Wei <654123789sw@gmail.com>

Commit:
kernel: add nft socket match support

removed nf_socket_ipv4 and nf_socket_ipv6 from the kmod-ipt-tproxy package
and added them to the kmod-nf-tproxy package

add nft_socket to the kmod-nft-tproxy package

Signed-off-by: Shen Wei <654123789sw@gmail.com>

Commit:
kernel: separat out the nft-socket package and separat ipv4 and ipv6 packets

separate the nft-socket package from the nft-tproxy package, add the
nft-socket package and make nft-socket dependent on nft-socket

separate the nf-defrag package from the nf-conntrack package and make
nf-conntrack dependent on nf-defrag

remove nf-tproxy dependency and tuning nft-tproxy depends on nft-core,
nf-defrag, nf-tproxy packages

separate nf-defrag, nf-socket, nf-tproxy all ipv4 and ipv6 packets

Signed-off-by: Shen Wei <654123789sw@gmail.com>

tiagogaspar8 commented on include/netfilter.mk:
```diff
@@ -152,7 +152,7 @@ $(eval $(if $(NF_KMOD),$(call nf_add,NF_REJECT6,CONFIG_NF_REJECT_IPV6, $(P_V6)nf
 $(eval $(if $(NF_KMOD),$(call nf_add,NF_IPT6,CONFIG_IP6_NF_IPTABLES, $(P_V6)ip6_tables),))
 
-$(eval $(if $(NF_KMOD),$(call nf_add,NF_CONNTRACK,CONFIG_NF_DEFRAG_IPV6, $(P_V6)nf_defrag_ipv6),))
+$(eval $(if $(NF_KMOD),$(call nf_add,NF_DEFRAG,CONFIG_NF_DEFRAG_IPV6, $(P_V6)nf_defrag_ipv6),))
```
Here I believe you forgot to add the "6"

shenwii replied:
Actually it is not the case, at first I correctly put nf_defrag_ipv4 into nf-defrag package and nf_defrag_ipv6 into nf-defrag6 package, then let nf_conntrack depend on nf-defrag and nf-defrag6, and as a result many other packages reported the dependency error. So I chose to make the same mistake as nf-conntrack and put nf_defrag_ipv6 into the nf-defrag package, while nf-defrag6 is also kept empty with nf-conntrack6.

> > I have a question about IPv6, originally nf-conntrack packet contains nf_defrag and nf_defrag6, but nf-conntrack6 packet is empty, why is that?

> That is indeed a mistake, that must be corrected.

Of course I only just learned that nf-conntrack is a mistake.

I think I need to fix the nf-conntrack bug together to make it work.

tiagogaspar8 replied:
Yeah, please give it a try, if you can't fix the dependency issue just do as conntrack and later I'll correct it if necessary.

tiagogaspar8 commented on package/kernel/linux/modules/netfilter.mk:
```diff
@@ -1190,3 +1253,25 @@ define KernelPackage/nft-compat
+  KCONFIG:=$(KCONFIG_NFT_TPROXY)
+endef
+
+$(eval $(call KernelPackage,nft-tproxy))
```
I think you should separate tproxy and socket into ipv4 and ipv6 versions.

shenwii replied:
```
config NFT_TPROXY
	tristate "Netfilter nf_tables tproxy support"
	depends on IPV6 || IPV6=n
	select NF_DEFRAG_IPV4
	select NF_DEFRAG_IPV6 if NF_TABLES_IPV6
	select NF_TPROXY_IPV4
	select NF_TPROXY_IPV6 if NF_TABLES_IPV6
	help
	  This makes transparent proxy support available in nftables.
```
Only nft-tproxy exists in the kernel, not nft-tproxy-ipv4 and nft-tproxy-ipv6, and even then do we have to separate nft-tproxy and nft-tproxy6?
Of course, nft-socket is too.

tiagogaspar8 replied:
Yeah, but if you see there is a line `select NF_TPROXY_IPV6 if NF_TABLES_IPV6` that selects NF_TPROXY_IPV6 if IPv6 is enabled, either you could do something like this +@IPV6:NF_TPROXY_IPV6 or you can separate them, choose whichever you think is correct.
Someday I'll have to standardize the netfilter config makefile, but only for nftables/when iptables is removed as this is a complete mess

shenwii replied:
Once I added `+@IPV6:` to the dependencies, the other package dependencies were fine.

Thanks again for the help given.

Commit:
kernel: fix ipv6 dependency of nft-socket related packages

put nf_defrag_ipv6 into the nf-defrag6 package correctly and make
nf-conntrack dependent on nf-defrag6

fix ipv6 dependency of nft-tproxy and nft-socket packages

Signed-off-by: Shen Wei <654123789sw@gmail.com>

tiagogaspar8 reviewed (approved):
Well, now I believe this issue is ready for the Openwrt maintainers with access to review this.
Great work!

hauke commented:
Similar patches were committed in 4f443c885dede3331b969e6265a41a0ff1e3059a, 0225df1ec8570d144a0a8d7862036c532cc44dbd and 43276b60c6dd5adeec89401c6ddd514e2cd22896.
Could you please check if all your features are working with them and if they look ok.

shenwii commented:
@tiagogaspar8 
That's a shame, what do I do next?
Should I turn this pull off and rebase the latest openwrt code and repull the nf-socket and nf-defrag code?

@hauke 
Thans, I will test it this evening.

tiagogaspar8 commented:
@shenwii Well, I'd say you should try the changes merged upstream, if they still don't work as you need/expect or bleive something is wrong/missing rebase this pull request and rename it to the new need
